### PR TITLE
Add up/down/pull commands

### PR DESCRIPTION
**Pull request by eugencowie**
### What does this change?

Initial design and structure for adding docker-compose commands as first-class fields. This PR includes pull/up/down as proof of concept, but this can easily be extended to other commands in a follow-on PR.

These commands use the same YAML syntax as the main `docker-compose` field for defining arguments and flags. For example:

```yaml
docker-compose:
  description: Pull services
  flags:
    file: compose.yml
  pull:
    arguments:
      - serviceA
      - serviceB
    flags:
      ignore-pull-failures:
```

This is equivalent to `docker-compose --file compose.yml pull --ignore-pull-failures serviceA serviceB`.

However, custom fields for each command could be added in the future (with appropriate defaults) without breaking backwards compatibility, allowing for a more concise and natural syntax as suggested in https://github.com/getporter/docker-compose-mixin/issues/3#issuecomment-886222995

### What issue does it fix?

Partially implements #3, though more work is needed to flesh it out.

### Notes for the reviewer

This is not a breaking change, the existing mixin syntax can still be used and works as expected.

I've updated the tests and examples to use the new syntax.

### Checklist

- [x] Did you write tests?
- [x] Did you write documentation?
- [x] Did you change porter.yaml or a storage document record? Update the corresponding schema file.
- [ ] If this is your first pull request, please add your name to the bottom of our [Contributors][contributors] list. Thank you for making Porter better! 🙇‍♀️

[contributors]: https://porter.sh/src/CONTRIBUTORS.md

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -169,18 +165,67 @@ docker-compose:
...
+  pull:
+    arguments:
+      - serviceA
+      - serviceB
+    flags:
+      ignore-pull-failures:
+      policy: missing
 ```
 
+This is equivilent to: `docker-compose --file compose.yml pull --ignore-pull-failures --policy missing serviceA serviceB`
```
Comment (kichristensen) [inline README.md:198]:
```suggestion
This is equivalent to: `docker-compose --file compose.yml pull --ignore-pull-failures --policy missing serviceA serviceB`
```

File: README.md [hunk 2]
```diff
@@ -169,18 +165,67 @@ docker-compose:
...
+  up:
+    arguments:
+      - serviceA
+      - serviceB
+    flags:
+      detach:
+      timeout: 30
+```
+
+This is equivilent to: `docker-compose up --detach --timeout 30 serviceA serviceB`
```
Comment (kichristensen) [inline README.md:214]:
```suggestion
This is equivalent to: `docker-compose up --detach --timeout 30 serviceA serviceB`
```

File: README.md [hunk 3]
```diff
@@ -169,18 +165,67 @@ docker-compose:
...
+```yaml
+docker-compose:
+  description: Docker Compose Down
+  down:
+    flags:
+      remove-orphans:
+      timeout: 30
+```
+
+This is equivilent to: `docker-compose down --remove-orphans --timeout 30`
```
Comment (kichristensen) [inline README.md:227]:
```suggestion
This is equivalent to: `docker-compose down --remove-orphans --timeout 30`
```

**Review by kichristensen — CHANGES_REQUESTED** [review]
Thanks for the PR, I will take a close look soon. But I have added a few small commets

**Comment by sgettys** [comment]
We should probably be doing the compose pull during build instead of action runtime so that all of the image dependencies are built into the porter bundle. I know this requires DinD so I need to see if it expects the images to be in the host systems docker cache or if it can use the bundle images docker cache to see if it would even honor pulling images at build time

**Comment by eugencowie** [comment]
> We should probably be doing the compose pull during build instead of action runtime so that all of the image dependencies are built into the porter bundle. I know this requires DinD so I need to see if it expects the images to be in the host systems docker cache or if it can use the bundle images docker cache to see if it would even honor pulling images at build time

Would that be resolved by #26?

**Comment by sgettys** [comment]
> > We should probably be doing the compose pull during build instead of action runtime so that all of the image dependencies are built into the porter bundle. I know this requires DinD so I need to see if it expects the images to be in the host systems docker cache or if it can use the bundle images docker cache to see if it would even honor pulling images at build time
> 
> Would that be resolved by #26?

I think partially although the docker-compose mixin build should be collecting those image at bundle build time and it's not. That's a secondary issue that we should also be addressing. I think this PR is good for now but I'm going to open an issue specifically for the build as running a "docker compose pull" at bundle runtime kind of defeats the purpose of CNAB.

**Review by kichristensen — APPROVED** [review]

**Pull request merged**